### PR TITLE
Add support for NativeScript Plugins

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
Add support for NativeScript plugins and npm packages. In order to use the packages you can use the following commands:
* `appbuilder plugin add <pluginName>@<pluginVersion>` (`@<pluginVersion>` is not mandatory to be used) - this command adds the plugin to package.json. If a path to local folder is provided, we check if the directory is part of the current project and if not - add it under plugins dir of the project. This way the plugin can be used.
* `appbuilder plugin remove <pluginName>` - removes the plugin from package.json and if it is part of the project - remove the source code (for example if it's been fetched)
* `appbuilder plugin fetch <pluginName>@<pluginVersion>` (`@<pluginVersion>` is not mandatory to be used) - gets the source code of the plugin inside plugins dir of the project and adds reference to it in package.json file. `<pluginName>` can be any github url, plugin published in npm, local directory, etc.
* `appbuilder plugin find <keyword1> <keyword2>...` - finds all plugins that match at least one of the keywords.
* `appbuilder plugin --available --count 20` - lists all available plugins. The count parameter defines number of Npm and NPM NativeScript plugins that will be shown on the screen.
* `appbuilder plugin add <github URL>` - adds the plugin to package.json (if the url is valid package) by using its name and set the url for "version", for example:
```
appbuilder plugin add https://github.com/rosen-vladimirov/NativeScriptPluginTest/tarball/master
```
should add to package.json:
```
"nativescript-plugin-test": "https://github.com/rosen-vladimirov/NativeScriptPluginTest/tarball/master"
```

In order to support plugins service for different type of projects, the following changes are applied:
* Add pluginsService property to IFrameworkProject. For CordovaProject it is resolve with "cordovaProjectPluginsService", for NS - "nativeScriptProjectPluginsService".
* Rename the previously used PluginsService to CordovaProjectPluginsService. The old file plugins-service had also been renamed to cordova-project-plugins-service.
* Add new pluginsService (new file plugins-service) that is used to call the project specific plugins service.
* Add nativescriptProjectPluginsService that operates with NativeScript plugins and NPM Packages.

>NOTE: In the future we can remove pluginsService and call directly `this.$project.pluginsService.<method>`, but currently this requires a lot of refactoring due to some cyclic dependencies.

Implements http://teampulse.telerik.com/view#item/298389

Merge with https://github.com/telerik/mobile-cli-lib/pull/464

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	"files.exclude": {
 		"**/.git": true,
 		"**/.DS_Store": true,
-		"**/*.js": { "when": "$(basename).ts"},
+		// "**/*.js": { "when": "$(basename).ts"},
```
Comment (rosen-vladimirov) [inline .vscode/settings.json]: undo this change

File: lib/commands/plugin/add-plugin.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ export class AddPluginCommand implements ICommand {
 								let defaultVersion = (<any>marketPlacePlugin.pluginVersionsData).DefaultVersion;
 
 								if(defaultVersion !== installedPlugin.data.Version) { // The default version is installed, we need to change DefaultVersion property
-									marketPlacePlugin.data = _.find(marketPlacePlugin.pluginVersionsData.Versions, versionData => versionData.Version === defaultVersion);
+									marketPlacePlugin.data = <any>_.find(marketPlacePlugin.pluginVersionsData.Versions, versionData => versionData.Version === defaultVersion);
```
Comment (rosen-vladimirov) [inline lib/commands/plugin/add-plugin.ts]: is this cast really needed ?

File: lib/commands/plugin/fetch-plugin.ts [hunk 3]
```diff
@@ -1,46 +1,17 @@
...
 
-import service = require("../../services/cordova-plugins");
-import validUrl = require("valid-url");
-
 export class FetchPluginCommand implements ICommand {
-	constructor(private $logger: ILogger,
-				private $fs: IFileSystem,
-				private $cordovaPluginsService: service.CordovaPluginsService,
-				private $stringParameterBuilder: IStringParameterBuilder) {
+	constructor(private $stringParameterBuilder: IStringParameterBuilder,
```
Comment (rosen-vladimirov) [inline lib/commands/plugin/fetch-plugin.ts]: use alphabetical order of constructor dependencies

File: lib/commands/plugin/find-plugins.ts [hunk 4]
```diff
@@ -1,25 +1,28 @@
 ///<reference path="../../.d.ts"/>
 "use strict";
 
-import * as service from "../../services/cordova-plugins";
 import {EOL} from "os";
 
 export class FindPluginsCommand implements ICommand {
-	constructor(private $cordovaPluginsService: service.CordovaPluginsService,
-				private $logger: ILogger) {
+	constructor(private $pluginsService: IPluginsService,
```
Comment (rosen-vladimirov) [inline lib/commands/plugin/find-plugins.ts]: use alphabetical order of constructor dependencies

File: lib/commands/plugin/find-plugins.ts [hunk 5]
```diff
@@ -1,25 +1,28 @@
...
+				private $logger: ILogger,
+				private $errors: IErrors) {
 	}
 
 	public allowedParameters: ICommandParameter[] = [];
 
 	public canExecute(args: string[]): IFuture<boolean> {
 		return (() => {
+			if(args.length === 0) {
```
Comment (rosen-vladimirov) [inline lib/commands/plugin/find-plugins.ts]: it can be just
```TypeScript
if(!args.length)
```

File: lib/declarations.d.ts [hunk 6]
```diff
@@ -523,7 +523,7 @@ interface ICordovaPluginsService {
 }
 
 interface IPluginsService {
-	getAvailablePlugins(): IPlugin[];
+	getAvailablePlugins(pluginsCount?: number): IPlugin[];
```
Comment (rosen-vladimirov) [inline lib/declarations.d.ts]: Add docs of the method

File: lib/declarations.d.ts [hunk 7]
```diff
@@ -543,16 +543,30 @@ interface IPluginsService {
...
 }
 
+interface IPluginInfoBase {
+	Authors: string[];
+	SupportedVersion: string;
+	Name: string;
+	Version: string;
+	Description: string;
+	Url: string;
+	// Platforms: Server.DevicePlatform[];
```
Comment (rosen-vladimirov) [inline lib/declarations.d.ts]: remove this

File: lib/declarations.d.ts [hunk 8]
```diff
@@ -608,6 +622,27 @@ interface IMarketplacePluginVersionsData extends Server.MarketplacePluginVersion
...
+ */
+interface IMarketplacePluginVersionsDataBase {
+	Versions: IMarketplacePluginData[];
+	Identifier: string;
+	DefaultVersion: string;
+	Framework: string;
+}
+
+/**
+ * Should be the same as Server's class
```
Comment (rosen-vladimirov) [inline lib/declarations.d.ts]: which one ?

File: lib/plugins-data.ts [hunk 9]
```diff
@@ -80,8 +80,8 @@ export class MarketplacePluginData extends CordovaPluginData {
 	private static TELERIK_PUBLISHER_NAME = "Telerik plugins";
 	private static TELERIK_PARTNER_PUBLISHER_NAME = "Telerik partner plugins";
 
-	constructor(public pluginVersionsData: IMarketplacePluginVersionsData,
-		public data: Server.MarketplacePluginData,
+	constructor(public pluginVersionsData: IMarketplacePluginVersionsDataBase,
```
Comment (rosen-vladimirov) [inline lib/plugins-data.ts]: use alphabetical order of constructor dependencies

File: lib/services/marketplace-plugins-service.ts [hunk 10]
```diff
@@ -12,8 +12,8 @@ export class MarketplacePluginsService implements ICordovaPluginsService {
 		return this.$server.cordova.getMarketplacePluginsData(this.$project.projectData.Framework);
 	}
 
-	public createPluginData(plugin: IMarketplacePluginVersionsData): IMarketplacePlugin[] { // DefaultVersion, Identifier, Versions
-		return _.map(plugin.Versions, (pluginVersionData) => new PluginsDataLib.MarketplacePluginData(plugin, pluginVersionData, this.$project, this.$projectConstants));
+	public createPluginData(plugin: IMarketplacePluginVersionsDataBase): IMarketplacePlugin[] { // DefaultVersion, Identifier, Versions
+		return <any[]>_.map(plugin.Versions, (pluginVersionData) => new PluginsDataLib.MarketplacePluginData(<any>plugin, <any>pluginVersionData, this.$project, this.$projectConstants));
```
Comment (rosen-vladimirov) [inline lib/services/marketplace-plugins-service.ts]: are all casts required?

File: lib/services/nativescript-project-plugins-service.ts [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+import * as shelljs from "shelljs";
+import * as semver from "semver";
+
+import PluginsDataLib = require("../plugins-data");
+
+export class NativeScriptProjectPluginsService implements IPluginsService {
+	private static MESSAGES = ["NPM Packages", "NPM NativeScript Plugins", "Marketplace Plugins"];
+
+	constructor(
+		private $errors: IErrors,
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: use alphabetical order of constructor dependencies

File: lib/services/nativescript-project-plugins-service.ts [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+		private $project: Project.IProject,
+		private $projectConstants: Project.IProjectConstants,
+		private $fs: IFileSystem,
+		private $server: Server.IServer,
+		private $httpClient: Server.IHttpClient,
+		private $childProcess: IChildProcess) {
+		}
+
+	public getAvailablePlugins(pluginsCount?: number): IPlugin[] {
+		let count = pluginsCount || 10;
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: 10 is a magic constant, move it to static field

File: lib/services/nativescript-project-plugins-service.ts [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+		let futures = [
+			this.getUniqueMarketplacePlugins(),
+			this.getTopNpmPackages(count),
+			this.getTopNativeScriptNpmPackages(count)
+		];
+		Future.wait(futures);
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: if one of the futures throws exception, the user will not receive any results. Is this expected? For example if npm is down, we will not show information for marketplace plugins as well.

File: lib/services/nativescript-project-plugins-service.ts [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+
+				return items;
+			}
+		}
+
+		return null;
+	}
+
+	public printPlugins(plugins: IPlugin[]): void {
+		let groups = _.groupBy(plugins, (plugin: IPlugin) => plugin.type);
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: this method shares the same code with CordovaProjectPluginsService, consider moving it to PluginsService

File: lib/services/nativescript-project-plugins-service.ts [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+					selectedPlugin = _.find(marketPlacePlugins, pl => selectedVersion.toLowerCase() === pl.data.Version.toLowerCase());
+				}
+				selectedVersion = selectedPlugin.data.Version;
+
+				if(!semver.satisfies(this.$project.projectData.FrameworkVersion, selectedPlugin.data.SupportedVersion)) {
+					this.$errors.failWithoutHelp(`Plugin ${pluginName} requires at least version ${selectedPlugin.data.SupportedVersion}, but your project targets ${this.$project.projectData.FrameworkVersion}.`);
+				}
+			} else {
+				try {
+					let url = `https://registry.npmjs.org/${pluginBasicInfo.name}?version=${selectedVersion}`;
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: `https://registry.npmjs.org` should be moved to a static field

File: lib/services/nativescript-project-plugins-service.ts [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+	public getPluginBasicInformation(pluginName: string): IBasicPluginInformation {
+		let [ name, version ] = pluginName.split("@");
+		return { name, version };
+	}
+
+	public fetch(pluginIdentifiers: string[]): IFuture<void> {
+		return (() => {
+			_.each(pluginIdentifiers, pluginId => {
+				try {
+					let pathToNodeModules = path.join(this.$project.getProjectDir().wait(), "node_modules");
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: `node_modules` should be moved to static field

File: lib/services/nativescript-project-plugins-service.ts [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+				}
+
+				// This call will return error with message '{}' in case there's no such package.
+				let result = this.$httpClient.httpRequest(url).wait().body;
+				let jsonResult = JSON.parse(result);
+				let platforms: string[];
+				let supportedVersion: string;
+				if(jsonResult.nativescript && jsonResult.nativescript.platforms) {
+					platforms = _.keys(jsonResult.nativescript.platforms);
+					supportedVersion = ">=" + _.first(_.values(jsonResult.nativescript.platforms));
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]:
```TypeScript
_.first(_.values(jsonResult.nativescript.platforms));
```
may not be correct with the following package.json
```
nativescript: {
      platforms: {
              ios: "1.1.0",
              android: "1.2.0"
     }
}
```

File: lib/services/nativescript-project-plugins-service.ts [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+
+	private installLocalPlugin(pluginPath: string): IFuture<void> {
+		return (() => {
+			let pathToPlugin = path.resolve(pluginPath);
+			let content = this.$fs.readJson(path.join(pathToPlugin, this.$projectConstants.PACKAGE_JSON_NAME)).wait();
+			let name = content.name;
+
+			if(pathToPlugin.indexOf(this.$project.getProjectDir().wait()) === -1) {
+				let pathToInstall = path.join(this.$project.getProjectDir().wait(), "plugins");
+				this.$logger.printMarkdown(`Moving \`${pathToPlugin}\` directory to \`${pathToInstall}\` in order to be able to use the plugin in your project.`);
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: consider using `util.format` in order to make the message easier for reading

File: lib/services/nativescript-project-plugins-service.ts [hunk 19]
```diff
@@ -0,0 +1,463 @@
...
+		}).future<boolean>()();
+	}
+
+	private installLocalPlugin(pluginPath: string): IFuture<void> {
+		return (() => {
+			let pathToPlugin = path.resolve(pluginPath);
+			let content = this.$fs.readJson(path.join(pathToPlugin, this.$projectConstants.PACKAGE_JSON_NAME)).wait();
+			let name = content.name;
+
+			if(pathToPlugin.indexOf(this.$project.getProjectDir().wait()) === -1) {
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: add comment why you need this check

File: lib/services/nativescript-project-plugins-service.ts [hunk 20]
```diff
@@ -0,0 +1,463 @@
...
+	public get pluginInformation(): string[] {
+		let additionalPluginData: string[];
+		if(this.data.Platforms && this.data.Platforms.length > 0) {
+			additionalPluginData = [this.buildRow("Platforms", this.data.Platforms.join(", "))];
+		}
+		return this.composePluginInformation(additionalPluginData);
+	}
+
+	public toProjectDataRecord(version: string): string {
+		return this.data.Name;
```
Comment (rosen-vladimirov) [inline lib/services/nativescript-project-plugins-service.ts]: is this correct? The plugin will not be added in the project (`.abproject`) at all

File: lib/services/plugins-service.ts [hunk 21]
```diff
@@ -1,643 +1,83 @@
...
-		}
-		return plugins;
+	getInstalledPlugins(): IPlugin[] {
+		return this.getPluginsService().wait().getInstalledPlugins();
 	}
-
-	private get configurations(): string[]{
-		return [this.$projectConstants.DEBUG_CONFIGURATION_NAME, this.$projectConstants.RELEASE_CONFIGURATION_NAME];
+	printPlugins(plugins: IPlugin[]): void {
+		// TODO: Implement
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: remove this comment

File: lib/services/plugins-service.ts [hunk 22]
```diff
@@ -1,643 +1,83 @@
...
-	}
-
-	private promptForVersion(pluginName: string, versions: any[]): IFuture<string> {
-		return (() => {
-			return versions.length > 1 ? this.promptForVersionCore(pluginName, versions).wait() : versions[0].value;
-		}).future<string>()();
-	}
-
-	private promptForVersionCore(pluginName: string, versions: any[]): IFuture<string> {
+	/**
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: remove these lines

File: lib/services/plugins-service.ts [hunk 23]
```diff
@@ -1,643 +1,83 @@
...
-					let currentVersionIndex = _.findIndex(versions,(v) => v.value === plugin.data.Version);
-					versions.splice(currentVersionIndex, 1);
-				}
-				version = this.promptForVersion(pluginName, versions).wait();
-			}
-
-			return version;
-		}).future<string>()();
+	isPluginInstalled(pluginName: string): boolean {
+		// TODO: Implement
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: remove this comment

File: lib/services/plugins-service.ts [hunk 24]
```diff
@@ -1,643 +1,83 @@
...
-					generalVariableValue = propValue;
-				}
-			});
-
-			let value = configVariableValue || generalVariableValue;
-			if(value) {
-				let obj = Object.create(null);
-				obj[variableName] = value.toString();
-				return obj;
+	private getPluginsService(): IFuture<IPluginsService> {
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: move this method below all public methods

File: lib/commands/plugin/add-plugin.ts (hunk 2 above)
Comment (rosen-vladimirov) [inline lib/commands/plugin/add-plugin.ts]: yes, because Versions collection is of type `Server.MarketplacePluginData`, while data property if of type `IPluginInfoBase`

File: lib/services/marketplace-plugins-service.ts (hunk 10 above)
Comment (rosen-vladimirov) [inline lib/services/marketplace-plugins-service.ts]: looks like yes, TypeScript is complaining about everything :crying_cat_face:

File: lib/commands/plugin/fetch-plugin.ts [hunk 25]
```diff
@@ -1,46 +1,14 @@
...
-	constructor(private $logger: ILogger,
-				private $fs: IFileSystem,
-				private $cordovaPluginsService: service.CordovaPluginsService,
-				private $stringParameterBuilder: IStringParameterBuilder) {
-	}
+	constructor(private $pluginsService: IPluginsService,
+				private $stringParameterBuilder: IStringParameterBuilder) {}
 
-	public allowedParameters = [this.$stringParameterBuilder.createMandatoryParameter("You must specify local path, URL to a plugin repository, name or keywords of a plugin published to the Cordova Plugin Registry.")];
+	public allowedParameters = [this.$stringParameterBuilder.createMandatoryParameter("You must specify local path, URL to a plugin repository, name or keywords of a plugin published to the NPM.")];
```
Comment (Mitko-Kerezov) [inline lib/commands/plugin/fetch-plugin.ts]: Consider changing this message, as we do not support npm packages in hybrid projects yet

File: lib/declarations.d.ts [hunk 26]
```diff
@@ -543,16 +586,75 @@ interface IPluginsService {
 	 * @return {IBasicPluginInformation}            Basic information about the plugin
 	 */
 	getPluginBasicInformation(pluginName: string): IBasicPluginInformation;
+
+	/**
+	 * Copies the source code of a plugin inside the project and adds it as a reference, so it can be used within the application.
+	 * @param {string[]} pluginIdentifiers The identifiers of the plugin that will be copied to the source code.
```
Comment (Mitko-Kerezov) [inline lib/declarations.d.ts]: small typo - `The identifiers of the plugin that will be copied to the source code.` should be something like `The identifiers of the plugins that will be copied to the source code.`

File: lib/declarations.d.ts [hunk 27]
```diff
@@ -608,6 +710,62 @@ interface IMarketplacePluginVersionsData extends Server.MarketplacePluginVersion
...
+	DownloadsCount: number;
+
+	/**
+	 * Assets
+	 */
+	Assets?: string[];
+
+	/**
+	 * Permissions that has to be approved by user when the application is used on Android
+	 * in order to use th plugin.
```
Comment (Mitko-Kerezov) [inline lib/declarations.d.ts]: in order to use th**e** plugin

File: lib/declarations.d.ts [hunk 28]
```diff
@@ -608,6 +710,62 @@ interface IMarketplacePluginVersionsData extends Server.MarketplacePluginVersion
...
+	 */
+	AndroidRequiredPermissions?: string[];
+
+	/**
+	 * Variables, that has to be set in the application in order to be able to use the plugin correctly.
```
Comment (Mitko-Kerezov) [inline lib/declarations.d.ts]: Variables, that **have** to be set...

File: lib/project/web-site-project.ts [hunk 29]
```diff
@@ -16,6 +16,7 @@ export class MobileWebSiteProject extends frameworkProjectBaseLib.FrameworkProje
 		super($logger, $fs, $resources, $errors, $jsonSchemaValidator, $options);
 	}
 
+	public pluginsService: IPluginsService;
```
Comment (Mitko-Kerezov) [inline lib/project/web-site-project.ts]: Not afraid of a null reference exception? :smiley_cat:

File: lib/services/cordova-project-plugins-service.ts [hunk 30]
```diff
@@ -0,0 +1,682 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+import {EOL} from "os";
+import * as util from "util";
+import * as pluginsDataLib from "../plugins-data";
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: As you are only using `PluginType` from `plugins-data` it can be captured as follows:
```Typescript
import {PluginType} from "../plugins-data";
```

File: lib/services/cordova-project-plugins-service.ts [hunk 31]
```diff
@@ -0,0 +1,682 @@
...
+import Future = require("fibers/future");
+import * as helpers from "../common/helpers";
+import semver = require("semver");
+import * as cordovaPlugins from "./cordova-plugins";
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: As above, this can be rewritten as
```Typescript
import {CordovaPluginsService} from "./cordova-plugins";

```

File: lib/services/cordova-project-plugins-service.ts [hunk 32]
```diff
@@ -0,0 +1,682 @@
...
+import * as validUrl from "valid-url";
+
+export class CordovaProjectPluginsService implements IPluginsService {
+	private static CORE_PLUGINS_PROPERTY_NAME = "CorePlugins";
+	private static CORDOVA_PLUGIN_VARIABLES_PROPERTY_NAME = "CordovaPluginVariables";
+	private static MESSAGES = ["Core Plugins", "Advanced Plugins", "Marketplace Plugins"];
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: Not a merge-stopper, however I do prefer the name `HEADINGS` over the name `MESSAGES` - again, personal opinion

File: lib/services/cordova-project-plugins-service.ts [hunk 33]
```diff
@@ -0,0 +1,682 @@
...
+		return this._identifierToPlugin;
+	}
+
+	private pluginsForbiddenConfigurations: IStringDictionary = {
+		"com.telerik.LivePatch": this.$projectConstants.DEBUG_CONFIGURATION_NAME
+	};
+
+	constructor(private $cordovaPluginsService: cordovaPlugins.CordovaPluginsService,
+		private $marketplacePluginsService: ICordovaPluginsService,
+		private $errors: IErrors,
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: use alphabetical order of constructor dependencies :smile_cat:

File: lib/commands/plugin/fetch-plugin.ts (hunk 25 above)
Comment (Mitko-Kerezov) [inline lib/commands/plugin/fetch-plugin.ts]: We've talked about this offline but I'm leaving a comment so that it won't be forgotten.
This parameter can be made non-mandatory and each of the plugin-services can handle validation on their own. That way each service can have a unique message.

File: lib/services/cordova-project-plugins-service.ts [hunk 34]
```diff
@@ -0,0 +1,682 @@
...
+		}).future<void>()();
+	}
+
+	public findPlugins(keywords: string[]): IFuture<IBasicPluginInformation[]> {
+		return (() => {
+			return this.$cordovaPluginsService.getPlugins(keywords);
+		}).future<IBasicPluginInformation[]>()();
+	}
+
+	public fetch(pluginIdentifiers: string[]): IFuture<void> {
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: This method was never intended to be used with multiple pluginIdentifiers. 
You can consider either changing the method's signature to accept only **one** pluginIdentifier, or refactor the method to actually be able to install multiple plugins.
My personal choice would be the first option as the second option would require more testing and refactoring for a functionality we do not yet need.

File: lib/services/cordova-project-plugins-service.ts [hunk 35]
```diff
@@ -0,0 +1,682 @@
...
+
+	private isLocalPath(pluginId: string): IFuture<boolean> {
+		return this.$fs.exists(pluginId);
+	}
+
+	private isUrlToRepository(pluginId: string): boolean {
+		return validUrl.isUri(pluginId);
+	}
+
+	public getInstalledPlugins(): IPlugin[] {
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: You can consider passing `debug` and `release` as `boolean` parameters instead of resolving them from `this.$options`

File: lib/services/cordova-project-plugins-service.ts [hunk 36]
```diff
@@ -0,0 +1,682 @@
...
+		});
+
+		if(!plugin) {
+			this.$errors.fail("Invalid plugin name: %s", pluginName);
+		}
+
+		return plugin;
+	}
+
+	public getPluginVersions(plugin: IPlugin): IPluginVersion[] {
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: Not a merge-stopper, consider moving this public method above the private ones

File: lib/services/marketplace-plugins-service.ts (hunk 10 above)
Comment (Mitko-Kerezov) [inline lib/services/marketplace-plugins-service.ts]: If you change the method's return type (as suggested by the interface in fact) then no casting is required.
Case in point: 
```Typescript
public createPluginData(plugin: IMarketplacePluginVersionsDataBase): IPlugin[] { // DefaultVersion, Identifier, Versions
	return _.map(plugin.Versions, (pluginVersionData) => new PluginsDataLib.MarketplacePluginData(plugin, pluginVersionData, this.$project, this.$projectConstants));
}
```

File: lib/services/nativescript-project-plugins-service.ts [hunk 37]
```diff
@@ -0,0 +1,472 @@
...
+"use strict";
+
+import * as path from "path";
+import * as util from "util";
+import {EOL} from "os";
+import Future = require("fibers/future");
+import * as shelljs from "shelljs";
+import * as semver from "semver";
+
+import PluginsDataLib = require("../plugins-data");
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: Only the `MarketplacePluginData` class is used from this file and can be captured as follows:
```Typescript 
import {MarketplacePluginData} from "../plugins-data";
```

File: lib/services/nativescript-project-plugins-service.ts [hunk 38]
```diff
@@ -0,0 +1,472 @@
...
+
+export class NativeScriptProjectPluginsService implements IPluginsService {
+	private static MESSAGES = ["NPM Packages", "NPM NativeScript Plugins", "Marketplace Plugins"];
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: Another non-merge stopper, same as the previous occurrence of this variable - I'd rather it was named differently.

File: lib/services/nativescript-project-plugins-service.ts [hunk 39]
```diff
@@ -0,0 +1,472 @@
...
+					this.$errors.failWithoutHelp(errorMessage);
+				}
+			});
+		}).future<void>()();
+	}
+
+	private marketplacePlugins: IPlugin[];
+	private getMarketplacePlugins(): IFuture<IPlugin[]> {
+		return ((): IPlugin[] => {
+			if(!this.marketplacePlugins) {
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: A better check would be either `!this.marketplacePlugins || !this.marketplacePlugins.length` or just `!this.marketplacePlugins.length`, where should you choose the latter the private field has of course to be declared as `private marketplacePlugins: IPlugin[] = []`.
I am writing this out of concern that should `this.marketplacePlugins` wind up being an empty array then no subsequent attempt will be mate to retrieve the plugins

File: lib/services/cordova-project-plugins-service.ts (hunk 35 above)
Comment (rosen-vladimirov) [inline lib/services/cordova-project-plugins-service.ts]: this will be handled in a separate PR

File: lib/services/nativescript-project-plugins-service.ts [hunk 40]
```diff
@@ -0,0 +1,472 @@
...
+			return plugins;
+		}).future<IPlugin[]>()();
+	}
+
+	private getDataForNpmPackage(packageName: string, version?: string): IFuture<IPlugin> {
+		return ((): IPlugin => {
+			let plugin: IPlugin;
+			try {
+				let url = `${NativeScriptProjectPluginsService.NPM_REGISTRY_URL}/${packageName}`;
+				let selectedVersion = version || "latest";
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: You don't seem to need this variable because you only ever assign a version if the original `version` variable is present.

File: lib/services/nativescript-project-plugins-service.ts [hunk 41]
```diff
@@ -0,0 +1,472 @@
...
+				if(version) {
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]: This `if` statement seems to be obsolete now

File: lib/services/cordova-project-plugins-service.ts [hunk 42]
```diff
@@ -0,0 +1,682 @@
...
+					return pvd && gr.data.Version === pvd.DefaultVersion;
+				});
+				if(defaultData) {
+					output.push(defaultData);
+				}
+			});
+
+			this.printPluginsCore(output);
+		} else {
+			this.printPluginsCore(plugins);
```
Comment (Mitko-Kerezov) [inline lib/services/cordova-project-plugins-service.ts]: The `if...else` statement leads to the same thing - a variable holding the end result can be introduced and `this.printPluginsCore` can be called at the end, omitting the `else`

File: lib/services/nativescript-project-plugins-service.ts [hunk 43]
```diff
@@ -0,0 +1,548 @@
...
+
+	public findPlugins(keywords: string[]): IFuture<IBasicPluginInformation[]> {
+		return ((): IBasicPluginInformation[] => {
+			let findPluginByNameFutures = _.map(keywords, keyword => this.findPluginsByName(keyword));
+			return getFuturesResults<IBasicPluginInformation>(findPluginByNameFutures, plugins => !!plugins && !!plugins.length);
+		}).future<IBasicPluginInformation[]>()();
+	}
+
+	public getInstalledPlugins(): IPlugin[] {
+		let pathToPackageJson = path.join(this.$project.getProjectDir().wait(), this.$projectConstants.PACKAGE_JSON_NAME);
```
Comment (Mitko-Kerezov) [inline lib/services/nativescript-project-plugins-service.ts]:
```Typescript
let pathToPackageJson = this.getPathToProjectPackageJson().wait();
```

**Comment by Mitko-Kerezov** [comment]
:+1: for the code and +:100: for the hard work and effort!
Fix rest of the comments, squash and go
